### PR TITLE
Validation improvements

### PR DESCRIPTION
**Pull request by Spacerat**
This PR improves validation in a few ways:

- `pre_save_full_clean_handler` in meta.py triggers model validation any time a model is saved (including `Model.object.create`)
- Adds validators for some model fields: `ResourceIdentifier.resource_id`,  `BuildResult.uuid` and `Project.uuid`.
- Some regexes are shared between python/models, HTML templates and the clientside JS. These are moved into a new module `ide/utils/regexes.py` and are made available to the templates with `{% regexes.regex_name %}` and the clientside JS with the global object `REGEXES.regex_name`.

(depends on #297)

### CODE REVIEW

File: ide/models/files.py [hunk 1]
```diff
@@ -168,7 +168,7 @@ class Meta(S3File.Meta):
 
 class ResourceIdentifier(IdeModel):
     resource_file = models.ForeignKey(ResourceFile, related_name='identifiers')
-    resource_id = models.CharField(max_length=100)
+    resource_id = models.CharField(max_length=100, validators=regexes.validator('c_identifier', _("Invalid resource ID.")))
```
Comment (Katharine) [inline ide/models/files.py]: This isn't quite right; resource IDs are prefixed and so don't have the usual start-of-identifier rules.

File: ide/utils/regexes.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+class RegexHolder(object):
+    regex_dictionary = {
+        # Match major[.minor], where major and minor are numbers between 0 and 255 with no leading 0s
+        'sdk_version': r'^(0|[1-9]\d?|1\d{2}|2[0-4]\d|25[0-5])(\.(0|[1-9]\d?|1\d{2}|2[0-4]\d|25[0-5]))?$',
+
+        # Match major.minor.0, where major and minor are numbers between 0 and 255 with no leading 0s
+        'semver': r'^(0|[1-9]\d?|1\d{2}|2[0-4]\d|25[0-5])\.(0|[1-9]\d?|1\d{2}|2[0-4]\d|25[0-5])\.0$',
+
+        # Match a string of letters and numbers separated with underscores but not starting with a digit
+        'c_identifier': r'^([a-zA-Z_][_a-zA-Z\d]*)$',
```
Comment (Katharine) [inline ide/utils/regexes.py]: More broadly, every use of this has the same property: it has a prefix when used in C, and so actually can start with a digit.

Comment (Spacerat) [inline ide/utils/regexes.py]: Good thing I can change every usage in one place! I suppose I could actually just use `^\w*$`.

Comment (Katharine) [inline ide/utils/regexes.py]: Don't forget `c_identifier_with_index`, which is the same thing with an added suffix. ;)

File: ide/models/files.py (hunk 1 above)
Comment (Spacerat) [inline ide/models/files.py]: Fixed

**Pull request merged**